localstack/services/ec2/resource_providers/aws_ec2_instance.py

```diff
@@ -0,0 +1,304 @@
+          - /properties/AvailabilityZone
+          - /properties/SubnetId
+          - /properties/LaunchTemplate
+          - /properties/SecurityGroups
+          - /properties/Ipv6AddressCount
+
+        Read-only properties:
+          - /properties/PublicIp
+          - /properties/Id
+          - /properties/PublicDnsName
+          - /properties/PrivateDnsName
+          - /properties/PrivateIp
+
+
+
+        """
+        model = request.desired_state
+        ec2 = request.aws_client_factory.ec2
+        # TODO: validations
+
+        if not request.custom_context.get(REPEATED_INVOCATION):
+            # this is the first time this callback is invoked
+            # TODO: defaults
+            # TODO: idempotency
+            params = util.select_attributes(
+                model,
+                ["InstanceType", "SecurityGroups", "KeyName", "ImageId", "MaxCount", "MinCount"],
+            )
+
+            if model.get("UserData"):
+                model["UserData"] = to_str(base64.b64decode(model["UserData"]))
+
+            response = ec2.run_instances(**params)
+            model["Id"] = response["Instances"][0]["InstanceId"]
+            request.custom_context[REPEATED_INVOCATION] = True
+            return ProgressEvent(
+                status=OperationStatus.IN_PROGRESS,
+                resource_model=model,
+                custom_context=request.custom_context,
+            )
+
+        response = ec2.describe_instances(InstanceIds=[model["Id"]])
+        instance = response["Reservations"][0]["Instances"][0]
+        if instance["State"]["Name"] != "running":
+            return ProgressEvent(
+                status=OperationStatus.IN_PROGRESS,
+                resource_model=model,
+                custom_context=request.custom_context,
+            )
+        model["PublicIp"] = instance["PublicIpAddress"]
+        model["PublicDnsName"] = instance["PublicDnsName"]
+        model["PrivateIp"] = instance["PrivateIpAddress"]
+        model["PrivateDnsName"] = instance["PrivateDnsName"]
+        model["AvailabilityZone"] = instance["Placement"]["AvailabilityZone"]
+
+        return ProgressEvent(
+            status=OperationStatus.SUCCESS,
+            resource_model=model,
+            custom_context=request.custom_context,
+        )
+
+    def read(
+        self,
+        request: ResourceRequest[EC2InstanceProperties],
+    ) -> ProgressEvent[EC2InstanceProperties]:
+        """
+        Fetch resource information
+
+
+        """
+        raise NotImplementedError
+
+    def delete(
+        self,
+        request: ResourceRequest[EC2InstanceProperties],
+    ) -> ProgressEvent[EC2InstanceProperties]:
+        """
+        Delete a resource
+
+
+        """
+        model = request.desired_state
+        ec2 = request.aws_client_factory.ec2
+        ec2.terminate_instances(InstanceIds=[model["Id"]])
+        # TODO add checking of ec2 instance state
+        return ProgressEvent(
```

Comment:
Normally this would be an async operation (i.e. requires use of the `REPEATED_INVOCATION` key) but in community this operation is in-memory only and does not create a VM, so it's fine.